OutOfSchool/OutOfSchool.Redis/MultiLayerCache.cs

```diff
@@ -0,0 +1,113 @@
+﻿using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OutOfSchool.Redis;
+
+public sealed class MultiLayerCache : IMultiLayerCacheService, IDisposable
+{
+    private readonly ReaderWriterLockSlim cacheLock = new ReaderWriterLockSlim();
+    private readonly IMemoryCache _memoryCache;
+    private readonly ICacheService _cacheService;
+    private readonly MemoryCacheConfig memoryCacheConfig;
+
+    public MultiLayerCache(
+        ICacheService externalCache,
+        IMemoryCache inMemoryCache,
+        IOptions<MemoryCacheConfig> memoryCacheConfig
+    )
+    {
+        this._memoryCache = inMemoryCache;
+        this._cacheService = externalCache;
+        this.memoryCacheConfig = memoryCacheConfig.Value;
+    }
+
+    public async Task RemoveAsync(string key)
```

Comment:
Excessive async/await here and in similar methods. Please check my comment in `CacheService.cs` for more details

Reply:
Deleted

Reply:
And it's still excessive. Please check the sharplab.io [snippet](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABADAAgwRgFYBuAWACgMBmPAJhwGEcBvCndnNj6vADjyQ4AggGcAngDswACgCUOALwA+PAE4cAWTnEOncroD0BnJAC2pmBIAuOAA6wAbgEsIAVxHDxUuTgCGEgBMcVykIc0sbUUkZeWAYABsIAHcAGhwrCBMACxgwAGscGAAzItybCCLChDAYEREnBxg+AwxVLnYjHgxBKO95ZU1tdr1dLsEteVZ9XW4AdgEAOgYw23iYKxgA7rJp9gBfCj2gA) to experiment what excessive async/await does with your code (you could there comment/uncomment one of two implementations of `Async()` method, pay attention to autogenerated `<Async>d__0` struct and imagine how it influences performance).

Here is simpler version with same functionality

```
    public Task RemoveAsync(string key)
    {
        _memoryCache.Remove(key);
        return _cacheService.RemoveAsync(key);
    }
```

Reply:
Done